OpenRA.Game/Primitives/Color.cs

```diff
@@ -117,9 +80,77 @@ public static Color FromArgb(uint argb)
 			return FromArgb((byte)(argb >> 24), (byte)(argb >> 16), (byte)(argb >> 8), (byte)argb);
 		}
 
+		static float SrgbToLinear(float c)
+		{
+			return c <= 0.04045f ? c / 12.92f : (float)Math.Pow((c + 0.055f) / 1.055f, 2.4f);
+		}
+
+		static float LinearToSrgb(float c)
+		{
+			return c < 0.0031308 ? c / 12.92f : ((float)Math.Pow(c, 5.0f / 12.0f) * 211.0f - 11.0f) / 200.0f;
+		}
+
+		public (float R, float G, float B) ToLinear()
+		{
+			// Undo pre-multiplied alpha and gamma correction
+			var r = SrgbToLinear((float)R / A);
+			var g = SrgbToLinear((float)G / A);
+			var b = SrgbToLinear((float)B / A);
+
+			return (r, g, b);
+		}
+
+		public static Color FromLinear(byte a, float r, float g, float b)
+		{
+			// Apply gamma correction and pre-multiplied alpha
+			return FromArgb(a,
+				(byte)Math.Round(LinearToSrgb(r) * a),
+				(byte)Math.Round(LinearToSrgb(g) * a),
+				(byte)Math.Round(LinearToSrgb(b) * a));
+		}
+
+		public static (float R, float G, float B) HsvToRgb(float h, float s, float v)
+		{
+			// Based on maths explained in http://lolengine.net/blog/2013/07/27/rgb-to-hsv-in-glsl
+			var px = Math.Abs(h * 6f - 3);
+			var py = Math.Abs((h + 2f / 3) % 1 * 6f - 3);
+			var pz = Math.Abs((h + 1f / 3) % 1 * 6f - 3);
+
+			var r = v * float2.Lerp(1f, (px - 1).Clamp(0, 1), s);
+			var g = v * float2.Lerp(1f, (py - 1).Clamp(0, 1), s);
+			var b = v * float2.Lerp(1f, (pz - 1).Clamp(0, 1), s);
+
+			return (r, g, b);
+		}
+
+		public static (float H, float S, float V) RgbToHsv(float r, float g, float b)
+		{
+			// Based on maths explained in http://lolengine.net/blog/2013/01/13/fast-rgb-to-hsv
+			var rgbMax = Math.Max(r, Math.Max(g, b));
+			var rgbMin = Math.Min(r, Math.Min(g, b));
+			var delta = rgbMax - rgbMin;
+
+			float hue;
+			if (r == rgbMax)
+				hue = (g - b) / (delta + 1e-20f);
+			else if (g == rgbMax)
+				hue = 2 + (b - r) / (delta + 1e-20f);
+			else
+				hue = 4 + (r - g) / (delta + 1e-20f);
+
+			if (hue < 0)
+				hue += 6;
+
+			var h = hue / 6;
+			var s = delta / (rgbMax + 1e-20f);
```

Comment:
I think you can do better than this: if delta is 0 you can basically assume hue and saturation are 0: you have a grey color, you just need to compute the value (brightness) and bail out. This will get rid of all the epsilon and simplify the code.

Reply:
Updated as discussed on Discord.